docs/RFCS/20190906_flow_control.md

```diff
@@ -0,0 +1,137 @@
+
+## Background
+
+Firstly we will discuss the key factor that affects the import or replication speed between DM data processing unit and downstream TiDB.
+
+- In the full data import procedure, data import speed is based on the SQL batch size and SQL executor concurrency. The SQL batch size is mainly determined by the single statement size of dump data, which can be controlled with the `mydumper` configuration. The SQL executor concurrency equals to the worker count of load unit, which is configurated by loader `pool-size` in task config.
+- In the incremental data replication procedure, replication speed relates to SQL job batch size and SQL executor concurrency. The binlog replication unit gets data from relay log consumer (assuming the relay log consume speed is fast enough), and distributes SQL jobs to multiple executors, the executor count is configurated by `worker-count` in task config. When the cached SQL job count exceeds SQL batch size (SQL batch size is configurated by `batch` in task config) or every 10ms, all SQLs in these jobs will be executed to downstream in a transaction.
+
+In the full data import scenario, it is easy to meet downstream congestion if we have many DM-workers and much high executor concurrency.
+
+While in incremental data replication scenario the congestion does not happen often, but there still exists some use scenario that may lead to congestion, including
+
+- User pauses replication task for a while and a lot of relay log is accumulated, and the replication node has high concurrency.
+- In shard scenario and there are too many upstream shard instances replicating to downstream at the same time.
+- In none shard scenario, but there are many tasks replicating to downstream from multiple instances at the same time.
+
+When we encounter these abnormal scenario, we often pause part of the tasks or decrease the SQL executor concurrency. The key point of solving the congestion problem is to reduce the data flow speed. Currently DM has no data flow control, we need to build a data flow control framework which will make data import or replication more smooth, more automotive and as fast as possible. The data flow control framework should provide the following features:
+
+- Downstream congestion definition and auto detection
+- Import or replication speed auto control
+
+## Implementation
+
+### Congestion definition
+
+There exists some key concepts in a data import/replication link:
+
+- Data loss: Each database transaction execution failure is treated as one data loss, the failure includes every kind of database execution failure that no data is written to downstream, partial data written error does not include.
+- Bandwidth: Bandwidth describes the maximum data import/replication rate from a DM-worker to downstream, this can be measured by tps from DM-worker (or TPS in downstream, but should only include the data traffic of this DM-worker).
```

Comment:
```suggestion
- Bandwidth: Bandwidth describes the maximum data import/replication rate from a DM-worker to downstream, this can be measured by TPS from DM-worker (or TPS in downstream, but should only include the data traffic from this DM-worker).
```